---
title: "GSE154881"
author: "Ben Barry"
date: "2025-01-26"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

```{r check-install}
pkg <- c("SummarizedExperiment", "extraChIPs", "edgeR", "tidyverse")
not_inst <- setdiff(pkg, rownames(installed.packages()))
if (length(not_inst)) BiocManager::install(not_inst)
```


```{r load-libraries}
library(tidyverse)
library(extraChIPs)
library(edgeR)
library(magrittr)
library(patchwork)
library(limma)
```


```{r loading data}
# create a count matrix
count_matrix <- read_tsv("/Users/benbarry/Documents/Telethon2025/data/raw/GSE154881_All_blood.txt") %>% 
  dplyr::filter(str_starts(Geneid, "ENSG")) %>% #keep all rows in GeneID column which start with "ENSG"
  as.data.frame() %>% # convert to data frame
  column_to_rownames("Geneid") %>% # convert "GeneID" to rownames
  as.matrix() # convert to matrix.
keep <- rowSums(count_matrix, na.rm = TRUE) > 0 & !is.na(rowSums(count_matrix)) # keep counts which are greater than 0 and not NA.

# save processed data
write.table(count_matrix, 
            file = "/Users/benbarry/Documents/Telethon2025/data/processed/count_matrix.txt")


#create a metadata table
GSE154881_metaData <- tibble(
  sampleName = colnames(count_matrix)
) %>% 
  mutate(
    disease_state = case_when(
      #make treatment "healthy_volunteer" if contains hv
      grepl("HV", sampleName) ~ "healthy_volunteer", 
      #make treatment "Diabetic_nephropathy" if contains dn
      grepl("DN", sampleName) ~ "Diabetic_nephropathy", 
      #make treatment "type2_diabetes" if contains dn
      grepl("T2D", sampleName) ~ "type2_diabetes" 
    ) %>%  
      factor(levels = c("healthy_volunteer", "type2_diabetes", "Diabetic_nephropathy"))
  )
#save meta data
write.table(GSE154881_metaData, 
            file = "/Users/benbarry/Documents/Telethon2025/data/processed/GSE154881_metaData.txt")


```

```{r annotating-ENSG}
# "Clean reads were aligned to the hg38 genome assembly using HISAT2 2.0.1 with default parameters.
# Raw counts of genes in different samples were calculated using featureCounts (version 2.0.0) based on the annotation gtf file.
# obtained from GENCODE version 23 (https://www.gencodegenes.org)"

library(rtracklayer)
library(plyranges)
GencodeV34 <- import.gff("/Users/benbarry/Documents/Telethon2025/data/raw/gencode.v34.chr_patch_hapl_scaff.annotation.gtf")

gene_lengths <- GencodeV34 %>% 
  subset(type == "exon") %>% 
  split(.$gene_id) %>% 
  reduce() %>% 
  width() %>% 
  map_int(sum)

genes_tbl <- GencodeV34 %>% 
  subset(type == "gene") %>% 
  as_tibble() %>% 
  dplyr::select(range, starts_with("gene"), -gene_status) %>% 
  mutate(gene_length = gene_lengths[gene_id])


rowRanges(se) <- GencodeV34 %>% 
  subset(type == "gene") %>% 
  select(starts_with("gene")) %>% 
  mutate(gene_length = gene_lengths[gene_id]) %>% 
  setNames(nm = mcols(.)$gene_id) %>% 
  .[rownames(se)]

setdiff(rownames(se), GencodeV34$gene_id)
setdiff(str_extract(rownames(se), "ENSG[0-9]+"), str_extract(GencodeV34$gene_id, "ENSG[0-9]+"))
```

```{r summarisedExperimentObject}
se <- SummarizedExperiment(
  assays = list(counts = count_matrix[keep,]), # create a summarisedExperiment object
  colData = GSE154881_metaData
)

#add a logCPM column to the assay
assay(se, "logCPM") <- cpm(se, log = TRUE)

# create a columns within rowData of se called detected where all values must bre greater than 1 and the row sum must be greater than or equal to 5
rowData(se)$detected <- se %>% 
  cpm() %>% 
  is_greater_than(1) %>% 
  rowSums() %>% 
  is_weakly_greater_than(5)
```

```{r plotting_with_ExraChIps}
se %>% 
  subset(detected) %>% 
  plotAssayDensities(assay = "logCPM", colour = "disease_state")
# seems like there are two donors who 
se %>% 
  subset(detected) %>% 
  plotAssayRle(assay = "logCPM", fill = "disease_state")

se %>% 
  subset(detected) %>% 
  plotAssayRle(assay = "logCPM", fill = "disease_state", by_x = "disease_state")

se %>%
  subset(detected) %>%
  plotAssayPCA(assay = "logCPM",
               colour =  "disease_state",
               pc_x = 1,
               pc_y = 2)

# based on density plot and RLE plot it seems like there are two Healthy Donors with poor quality samples

```



#TMM normalisation - calcNormFactors
# plot PCA + RLE 


```{r quick_subset_of_detected}
se_detected <- subset(se, detected) 

assay(se_detected, "norm_logCPM") <- se_detected %>% 
  calcNormFactors() %>% 
  cpm(log = TRUE)

se_detected %>%
  plotAssayPCA(assay = "norm_logCPM",
               colour =  "disease_state",
               label = "sampleName",
               pc_x = 1,
               pc_y = 2)


```

``` {r }
# creating a summarized experiment for detected data set

detectedCountMatrix <-se@assays@data@listData[["counts"]]

detectedCountMatrixMetaData <- tibble(
  sampleName = colnames(detectedCountMatrix)
) %>% 
  mutate(
    disease_state = case_when(
      #make treatment "healthy_volunteer" if contains hv
      grepl("HV", sampleName) ~ "healthy_volunteer", 
      #make treatment "Diabetic_nephropathy" if contains dn
      grepl("DN", sampleName) ~ "Diabetic_nephropathy", 
      #make treatment "type2_diabetes" if contains dn
      grepl("T2D", sampleName) ~ "type2_diabetes" 
    ) %>%  
      factor(levels = c("healthy_volunteer", "type2_diabetes", "Diabetic_nephropathy"))
  )  

DetectedSE <- SummarizedExperiment(
  assays = list(counts = detectedCountMatrix), # create a summarisedExperiment object
  colData = detectedCountMatrixMetaData 
)

#normalise By creating a DGEList and using calc norm factors
assay(DetectedSE, "TmmNormCPM") <- DGEList(counts = detectedCountMatrix) %>%
  calcNormFactors(method = "TMM") %>%
  cpm(log = TRUE)

DetectedSE %>% 
  plotAssayRle(assay = "TmmNormCPM", fill = "disease_state")

DetectedSE %>% 
  plotAssayRle(assay = "TmmNormCPM", fill = "disease_state", by_x = "disease_state")

DetectedSE %>% 
  plotAssayPCA(assay = "TmmNormCPM",
               colour =  "disease_state",
               label = "sampleName",
               pc_x = 1,
               pc_y = 2)

#Following TMM nomrmalisation there is still large varience in the data hence HV3 and HV4 samples will be removed.  
```

```{r} 
# option 1:
#remove Problematic samples.
SE_outlier_removed <- se_detected[, !se_detected$sampleName %in% c("HV_3", "HV_4")]

#calc normalisation factors using TMM
assay(SE_outlier_removed, "TmmNorm") <- SE_outlier_removed %>% 
  calcNormFactors(method = "TMM") %>% 
  cpm(log = TRUE)

SE_outlier_removed %>% 
  plotAssayRle(assay = "TmmNorm", fill = "disease_state")

SE_outlier_removed%>% 
  plotAssayRle(assay = "TmmNorm", fill = "disease_state", by_x = "disease_state")

A <- SE_outlier_removed %>% 
  plotAssayPCA(assay = "TmmNorm",
               colour =  "disease_state",
               label = "sampleName",
               pc_x = 1,
               pc_y = 2)  
B <- SE_outlier_removed %>% 
  plotAssayPCA(assay = "TmmNorm",
               colour =  "disease_state",
               label = "sampleName",
               pc_x = 2,
               pc_y = 3)  
A + B +
  plot_layout(guides = "collect", axes = "collect") &
  theme(legend.position = "bottom")

write_rds(SE_outlier_removed, "/Users/benbarry/Documents/Telethon2025/data/processed/SE_outlier_removed.rds", compress = "gz")

```


