<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ben Barry" />

<meta name="date" content="2025-03-25" />

<title>LoadTranformCountData</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/sandstone.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Type 2 diabetes analysis (GSE154881)</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Raw data QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="LoadTranformCountData.html">Intial load and data transformation</a>
    </li>
    <li>
      <a href="LowCountSampleRemoval.html">removal of low quality samples</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Differential Gene Expression Analysis (DGEA)
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="TmmDGEA.html">Tmm Normalised Data</a>
    </li>
    <li>
      <a href="RUVgNorm.html">RUVg Normalised Data</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    enrichment analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="GOseq.html">Goseq</a>
    </li>
    <li>
      <a href="GSEA.html">GSEA</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bdbarry/Telethon2025_RNAseq_Analysis">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">LoadTranformCountData</h1>
<h4 class="author">Ben Barry</h4>
<h4 class="date">2025-03-25</h4>

</div>


<pre class="r"><code>library(tidyverse)
library(extraChIPs)
library(edgeR)
library(magrittr)
library(patchwork)
library(limma)
library(here)
library(rtracklayer)
library(plyranges)</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>Raw count data was obtained from ‘GSE154881’. This data contains 15
samples from 3 disease states, healthy, type 2 Diabetic and Diabetic
Nephropathy. Methodological information can be found at the GEO site,
however, of note a RIN score of &gt;5 was used as the threshold for RNA
quality, this is rather low and may reflect downstream QC decisions.</p>
<div id="count-matrix" class="section level3">
<h3>count matrix</h3>
<pre class="r"><code># create a count matrix
count_matrix &lt;- read_tsv(here::here(&quot;data/raw/GSE154881_All_blood.txt&quot;)) %&gt;% 
  dplyr::filter(str_starts(Geneid, &quot;ENSG&quot;)) %&gt;% #keep all rows in GeneID column which start with &quot;ENSG&quot;
  as.data.frame() %&gt;% # convert to data frame
  column_to_rownames(&quot;Geneid&quot;) %&gt;% # convert &quot;GeneID&quot; to rownames
  as.matrix() # convert to matrix.
keep &lt;- rowSums(count_matrix, na.rm = TRUE) &gt; 0 &amp; !is.na(rowSums(count_matrix)) # keep counts which are greater than 0 and not NA.
head(count_matrix[keep,])</code></pre>
<pre><code>##                   T2D_1 T2D_2 T2D_3 T2D_4 T2D_5 DN_1 DN_2 DN_3 DN_4 DN_5 HV_1
## ENSG00000223972.5     2     5    11    10     7    3   17    8   10   36    1
## ENSG00000227232.5    99    67    33    49    61   86   29   13   65   27   43
## ENSG00000278267.1     0     0     1     1     8    0    6    0    1    4    2
## ENSG00000243485.5     0     0     0     0     0    0    1    0    0    0    0
## ENSG00000238009.6     0     1     0     0     0    3    1    3    0    2    6
## ENSG00000233750.3    48    19    21    16     3   18    4   12   20   19   54
##                   HV_2 HV_3 HV_4 HV_5
## ENSG00000223972.5    8   17    6   22
## ENSG00000227232.5   11   10    5    8
## ENSG00000278267.1    2    0    1    0
## ENSG00000243485.5    0    0    0    0
## ENSG00000238009.6    0    0    0    0
## ENSG00000233750.3   19    7    1    3</code></pre>
<p>Following download of data, an initial parameter was set to only keep
transcripts whose counts were greater than 0. A count matrix with 15
samples and 60669 alignments was saved locally.</p>
</div>
<div id="metadata" class="section level3">
<h3>metadata</h3>
<p>Meta data was also extracted and saved locally</p>
<pre class="r"><code>#create a metadata table
GSE154881_metaData &lt;- tibble(
  sampleName = colnames(count_matrix)
) %&gt;% 
  mutate(
    disease_state = case_when(
      #make treatment &quot;healthy_volunteer&quot; if contains hv
      grepl(&quot;HV&quot;, sampleName) ~ &quot;healthy_volunteer&quot;, 
      #make treatment &quot;Diabetic_nephropathy&quot; if contains dn
      grepl(&quot;DN&quot;, sampleName) ~ &quot;Diabetic_nephropathy&quot;, 
      #make treatment &quot;type2_diabetes&quot; if contains dn
      grepl(&quot;T2D&quot;, sampleName) ~ &quot;type2_diabetes&quot; 
    ) %&gt;%  
      factor(levels = c(&quot;healthy_volunteer&quot;, &quot;type2_diabetes&quot;, &quot;Diabetic_nephropathy&quot;))
  )

head(GSE154881_metaData)</code></pre>
<pre><code>## # A tibble: 6 × 2
##   sampleName disease_state       
##   &lt;chr&gt;      &lt;fct&gt;               
## 1 T2D_1      type2_diabetes      
## 2 T2D_2      type2_diabetes      
## 3 T2D_3      type2_diabetes      
## 4 T2D_4      type2_diabetes      
## 5 T2D_5      type2_diabetes      
## 6 DN_1       Diabetic_nephropathy</code></pre>
</div>
<div id="gene-name-annotation" class="section level3">
<h3>gene name annotation</h3>
<p>Despite the methods stating that raw counts were calculated using a
GTF file from GENCODEv23, debugging revealed that this was an error and
should have been GENCODEv34. Exon gene lengths were then calculated by
merging exon ranges which overlap and finding the total exon length.
This was performed for use in downstream analysis.</p>
<pre class="r"><code>GencodeV34 &lt;- import.gff(here::here(&quot;data/raw/gencode.v34.chr_patch_hapl_scaff.annotation.gtf&quot;))

gene_lengths &lt;- GencodeV34 %&gt;% 
  subset(type == &quot;exon&quot;) %&gt;%   # keep only exons
  split(.$gene_id) %&gt;%        # similar to &quot;group_by&quot; in tidyverse
  reduce() %&gt;%                # merges exon ranges that fall within others. eg. 2:5 and 1:7 becomes 1:7
  width() %&gt;%                 # calculate the length of the each of ranges
  map_int(sum)                # calculates the total exon length

gene_lengths[1:10]</code></pre>
<pre><code>## ENSG00000000003.15  ENSG00000000005.6 ENSG00000000419.12 ENSG00000000457.14 
##               4536               1476               1207               6883 
## ENSG00000000460.17 ENSG00000000938.13 ENSG00000000971.16 ENSG00000001036.14 
##               5970               3382               7980               2822 
## ENSG00000001084.13 ENSG00000001167.14 
##               8618               3811</code></pre>
</div>
<div id="summarised-object" class="section level3">
<h3>summarised object</h3>
<p>the data described thus far was compiled into a summarized experiment
(SE) object for ease of handling. Metadata was assigned to the colData
element of the the SE object and gene annotations assigned to the
RowRanges obejct. LogCPM were calculated and assinged to a new Assay
within the SE object. A detection parameter was also added where CPM
values aught to be greater than 1 and the row total &gt; 5</p>
<pre class="r"><code>se &lt;- SummarizedExperiment(                    # create a summarisedExperiment object
  assays = list(counts = count_matrix[keep,]), # add the countmatrix as an assay
  colData = GSE154881_metaData                 # add the metadata
)

# add gene annotations to SE object
rowRanges(se) &lt;- GencodeV34 %&gt;% 
  subset(type == &quot;gene&quot;) %&gt;%                      #keep only genes
  select(starts_with(&quot;gene&quot;)) %&gt;%                 #keep only the columns which starts with &quot;gene&quot;
  mutate(gene_length = gene_lengths[gene_id]) %&gt;% #create a new columns called gene_length using the vector from                                                         gene_lengths
  setNames(nm = mcols(.)$gene_id) %&gt;%             #set name using the column names from gene_id in gencodeV34 
  .[rownames(se)]                                 #reorder the GencodeV34 to match rownames in se

#add a logCPM column to the assay
assay(se, &quot;logCPM&quot;) &lt;- cpm(se, log = TRUE)

# create a columns within rowData of se called detected where all values must be greater than 1 and the row sum must be greater than or equal to 5
rowData(se)$detected &lt;- se %&gt;% 
  cpm() %&gt;% 
  is_greater_than(1) %&gt;% 
  rowSums() %&gt;% 
  is_weakly_greater_than(5)

glimpse(se)</code></pre>
<pre><code>## Formal class &#39;RangedSummarizedExperiment&#39; [package &quot;SummarizedExperiment&quot;] with 6 slots
##   ..@ rowRanges      :Formal class &#39;GRanges&#39; [package &quot;GenomicRanges&quot;] with 7 slots
##   ..@ colData        :Formal class &#39;DFrame&#39; [package &quot;S4Vectors&quot;] with 6 slots
##   ..@ assays         :Formal class &#39;SimpleAssays&#39; [package &quot;SummarizedExperiment&quot;] with 1 slot
##   ..@ NAMES          : NULL
##   ..@ elementMetadata:Formal class &#39;DFrame&#39; [package &quot;S4Vectors&quot;] with 6 slots
##   ..@ metadata       : list()</code></pre>
</div>
<div id="log-cpm-plots" class="section level3">
<h3>Log CPM plots</h3>
<p>Log CPM was then plotted using ‘ExtrChIps’ to observe variation
between donors and disease states. First we utilized
<code>plotAssayDensities</code> to visualize the kernel density
estimates - in essence this provides a smoothed histogram. Here we would
expect each sample to exhibit a similar distribution of read counts.</p>
<pre class="r"><code>library(RColorBrewer)
se %&gt;% 
  subset(detected) %&gt;% 
  plotAssayDensities(assay = &quot;logCPM&quot;, colour = &quot;disease_state&quot;, size = 1) + 
  theme_bw() +
  xlim(c(-5, 20)) +
  scale_colour_brewer(palette = &quot;Accent&quot;)</code></pre>
<p><img src="LoadTranformCountData_files/figure-html/Figure1:DensityPlot-1.png" width="960" /></p>
</div>
<div id="relative-log-expression" class="section level3">
<h3>Relative Log Expression</h3>
<p>From this density plot it appears that there are two samples from the
“healthy_volunteer” cohort which appear to have a negatively skewed
LogCPM, i.e. a high frequency of low counts. To further interrogate this
we next plotted the relative log expression (RLE). This performs a Log1p
transformation of the data and presents the results as a boxplot,
allowing for visualization of variance.</p>
<pre class="r"><code>se %&gt;% 
  subset(detected) %&gt;% 
  plotAssayRle(assay = &quot;logCPM&quot;, fill = &quot;disease_state&quot;) + 
  theme_bw() +
  scale_fill_brewer(palette = &quot;Accent&quot;)</code></pre>
<p><img src="LoadTranformCountData_files/figure-html/RlePlot-1.png" width="960" /></p>
<pre class="r"><code>se %&gt;% 
  subset(detected) %&gt;% 
  plotAssayRle(assay = &quot;logCPM&quot;, fill = &quot;disease_state&quot;, by_x = &quot;disease_state&quot;) +
  theme_bw() +
  scale_fill_brewer(palette = &quot;Accent&quot;)</code></pre>
<p><img src="LoadTranformCountData_files/figure-html/RlePlot-2.png" width="960" />
These plots illustrate the variance within these data, in particular,
highlighting samples 3 and 4 from the healthy volunteer cohort (HV_3,
HV_4) as exhibiting substantial variance.</p>
</div>
<div id="principal-component-analysis-pca" class="section level3">
<h3>Principal Component Analysis (PCA)</h3>
<p>To gain further insight into this observed variance, PCA was
performed on the LogCPM data to understand the variance which may be
contributed by HV_3 and HV_4.</p>
<pre class="r"><code>se %&gt;%
  subset(detected) %&gt;%
  plotAssayPCA(
    assay = &quot;logCPM&quot;, 
    colour =  &quot;disease_state&quot;, label = &quot;sample&quot;,
    pc_x = 1, pc_y = 2
  ) +
  geom_point(shape = 1,size = 2,colour = &quot;black&quot;) + 
  theme_bw() +
  scale_colour_brewer(palette = &quot;Accent&quot;)</code></pre>
<p><img src="LoadTranformCountData_files/figure-html/PCA-1.png" width="960" />
Here we observed that HV_3 and HV_4 segregate along PC1, the component
which explains 56% of the variance in the data. This indicates that the
these samples contribute the most to the observed variance in the
data.</p>
<p>Collectively these analysis provide a sound indication that these
samples may not be an accurate representation of the healthy volunteer
sample and therefor population. Although challenging to definitively
determine why this may be, the low RIN threshold may explain these low
counts.</p>
</div>
<div id="tmm-normalisation" class="section level3">
<h3>TMM Normalisation</h3>
<p>Next we explored whether this observed variance could be mitigated by
TMM normalization.</p>
<pre class="r"><code>se_detected &lt;- subset(se, detected) 

assay(se_detected, &quot;TMM_logCPM&quot;) &lt;- se_detected %&gt;% 
  calcNormFactors(&quot;TMM&quot;) %&gt;% 
  cpm(log = TRUE)

se_detected %&gt;%
  plotAssayRle(assay = &quot;TMM_logCPM&quot;, fill = &quot;disease_state&quot;) +
  theme_bw() +
  scale_fill_brewer(palette = &quot;Accent&quot;)</code></pre>
<p><img src="LoadTranformCountData_files/figure-html/TMMnorm-1.png" width="960" /></p>
<pre class="r"><code>se_detected %&gt;%
  plotAssayPCA(assay = &quot;TMM_logCPM&quot;,
               colour =  &quot;disease_state&quot;,
               label = &quot;sampleName&quot;,
               pc_x = 1,
               pc_y = 2) + 
  scale_colour_brewer(palette = &quot;Accent&quot;) +
  geom_point(shape = 1,size = 2,colour = &quot;black&quot;)</code></pre>
<p><img src="LoadTranformCountData_files/figure-html/TMMnorm-2.png" width="960" /></p>
<p>RLE plot reveals that HV_3 and HV_4 still exhibit relative large
variance (indicated by their mean deviating from 0) Similarly PCA
reveals these two samples still contribute to variance in the data, even
after normalization.</p>
<p>Ultimately this has led to the descion to exclude these two sample
from further analysis.</p>
<pre class="r"><code>se_detected &lt;- subset(se, detected) 
SE_outlier_removed &lt;- se_detected[, !se_detected$sampleName %in% c(&quot;HV_3&quot;, &quot;HV_4&quot;)]</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
